File: frontend/services/chat/chatContracts.ts
export type ChatRole =
  | "system"
  | "user"
  | "assistant"
  | "tool";

export type ThreadStatus =
  | "active"
  | "archived";

export type MemoryStatus =
  | "active"
  | "superseded"
  | "retracted";

export type EvidenceStatus =
  | "success"
  | "error"
  | "unavailable";

export type ToolTruthState =
  | "grounded"
  | "partial"
  | "ungrounded"
  | "not_required";

export type ChatThreadContract = {
  thread_id: string;
  user_id?: string | null;
  title?: string | null;
  status: ThreadStatus;
  summary?: string | null;
  message_count: number;
  created_at: string;
  updated_at: string;
  metadata: Record<string, unknown>;
};

export type ChatMessageContract = {
  message_id: string;
  thread_id: string;
  parent_message_id?: string | null;
  role: ChatRole;
  content: string;
  intent?: string | null;
  symbol?: string | null;
  provider?: string | null;
  model?: string | null;
  tool_evidence_ids: string[];
  created_at: string;
  metadata: Record<string, unknown>;
};

export type PendingConversationTaskContract = {
  task_id: string;
  task_type: string;
  description: string;
  status:
    | "pending"
    | "running"
    | "completed"
    | "failed"
    | "cancelled";
  source_message_id?: string | null;
  details: Record<string, unknown>;
};

export type ConversationStateContract = {
  thread_id: string;
  summary: string;
  active_entities: Record<string, unknown>;
  active_strategy?: Record<string, unknown> | null;
  portfolio_context: Record<string, unknown>;
  pending_tasks: PendingConversationTaskContract[];
  remembered_terms: Record<string, string>;
  last_message_id?: string | null;
  updated_at: string;
  metadata: Record<string, unknown>;
};

export type MemoryFactContract = {
  memory_id: string;
  thread_id: string;
  key: string;
  value: unknown;
  confidence: number;
  status: MemoryStatus;
  source_message_id?: string | null;
  created_at: string;
  updated_at: string;
  metadata: Record<string, unknown>;
};

export type ToolEvidenceContract = {
  evidence_id: string;
  thread_id: string;
  tool_name: string;
  tool_call_id?: string | null;
  claim_types: string[];
  request: Record<string, unknown>;
  result: Record<string, unknown>;
  status: EvidenceStatus;
  observed_at: string;
  expires_at?: string | null;
  metadata: Record<string, unknown>;
};

export type ThreadedChatRequest = {
  message: string;
  thread_id?: string | null;
  parent_message_id?: string | null;
  client_message_id?: string | null;
  stream?: boolean;
  metadata?: Record<string, unknown>;
};

export type ThreadedChatResponse = {
  status: "ok" | "error" | "blocked";
  thread_id: string;
  user_message_id: string;
  assistant_message_id?: string | null;
  message: string;
  intent?: string | null;
  symbol?: string | null;
  provider?: string | null;
  model?: string | null;
  tool_truth_state: ToolTruthState;
  evidence: ToolEvidenceContract[];
  memory_updates: MemoryFactContract[];
  conversation_state?: ConversationStateContract | null;
  error?: string | null;
  metadata: Record<string, unknown>;
};

export const CHAT_CONTRACT_VERSION =
  "134A.v1" as const;

File: backend/app/stacks/chat_public/contracts/chat_contracts.py
"""
134A_CHAT_THREAD_AND_MEMORY_CONTRACT

Canonical contracts for NeuroVest conversation threads, messages,
conversation state, memory facts, tool evidence, requests, and responses.

This phase defines structure only.

It does not:
- persist conversations
- alter Ollama prompts
- add streaming
- execute tools
- change broker or trading state
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any, Literal
from uuid import uuid4

from pydantic import BaseModel, Field


ChatRole = Literal[
    "system",
    "user",
    "assistant",
    "tool",
]

ThreadStatus = Literal[
    "active",
    "archived",
]

MemoryStatus = Literal[
    "active",
    "superseded",
    "retracted",
]

EvidenceStatus = Literal[
    "success",
    "error",
    "unavailable",
]

ToolTruthState = Literal[
    "grounded",
    "partial",
    "ungrounded",
    "not_required",
]


def utc_now() -> datetime:
    return datetime.now(UTC)


def new_id(prefix: str) -> str:
    return f"{prefix}-{uuid4().hex}"


class ChatThread(BaseModel):
    thread_id: str = Field(
        default_factory=lambda: new_id("thread")
    )

    user_id: str | None = None
    title: str | None = None

    status: ThreadStatus = "active"

    summary: str | None = None
    message_count: int = 0

    created_at: datetime = Field(
        default_factory=utc_now
    )

    updated_at: datetime = Field(
        default_factory=utc_now
    )

    metadata: dict[str, Any] = Field(
        default_factory=dict
    )


class ChatMessage(BaseModel):
    message_id: str = Field(
        default_factory=lambda: new_id("message")
    )

    thread_id: str

    parent_message_id: str | None = None

    role: ChatRole
    content: str

    intent: str | None = None
    symbol: str | None = None

    provider: str | None = None
    model: str | None = None

    tool_evidence_ids: list[str] = Field(
        default_factory=list
    )

    created_at: datetime = Field(
        default_factory=utc_now
    )

    metadata: dict[str, Any] = Field(
        default_factory=dict
    )


class PendingConversationTask(BaseModel):
    task_id: str = Field(
        default_factory=lambda: new_id("task")
    )

    task_type: str
    description: str

    status: Literal[
        "pending",
        "running",
        "completed",
        "failed",
        "cancelled",
    ] = "pending"

    source_message_id: str | None = None

    details: dict[str, Any] = Field(
        default_factory=dict
    )


class ConversationState(BaseModel):
    thread_id: str

    summary: str = ""

    active_entities: dict[str, Any] = Field(
        default_factory=dict
    )

    active_strategy: dict[str, Any] | None = None

    portfolio_context: dict[str, Any] = Field(
        default_factory=dict
    )

    pending_tasks: list[PendingConversationTask] = Field(
        default_factory=list
    )

    remembered_terms: dict[str, str] = Field(
        default_factory=dict
    )

    last_message_id: str | None = None

    updated_at: datetime = Field(
        default_factory=utc_now
    )

    metadata: dict[str, Any] = Field(
        default_factory=dict
    )


class MemoryFact(BaseModel):
    memory_id: str = Field(
        default_factory=lambda: new_id("memory")
    )

    thread_id: str

    key: str
    value: Any

    confidence: float = Field(
        default=1.0,
        ge=0.0,
        le=1.0,
    )

    status: MemoryStatus = "active"

    source_message_id: str | None = None

    created_at: datetime = Field(
        default_factory=utc_now
    )

    updated_at: datetime = Field(
        default_factory=utc_now
    )

    metadata: dict[str, Any] = Field(
        default_factory=dict
    )


class ToolEvidence(BaseModel):
    evidence_id: str = Field(
        default_factory=lambda: new_id("evidence")
    )

    thread_id: str

    tool_name: str
    tool_call_id: str | None = None

    claim_types: list[str] = Field(
        default_factory=list
    )

    request: dict[str, Any] = Field(
        default_factory=dict
    )

    result: dict[str, Any] = Field(
        default_factory=dict
    )

    status: EvidenceStatus

    observed_at: datetime = Field(
        default_factory=utc_now
    )

    expires_at: datetime | None = None

    metadata: dict[str, Any] = Field(
        default_factory=dict
    )


class ChatRequest(BaseModel):
    message: str = Field(
        min_length=1
    )

    thread_id: str | None = None
    parent_message_id: str | None = None
    client_message_id: str | None = None

    stream: bool = False

    metadata: dict[str, Any] = Field(
        default_factory=dict
    )


class ChatResponse(BaseModel):
    status: Literal[
        "ok",
        "error",
        "blocked",
    ]

    thread_id: str

    user_message_id: str
    assistant_message_id: str | None = None

    message: str

    intent: str | None = None
    symbol: str | None = None

    provider: str | None = None
    model: str | None = None

    tool_truth_state: ToolTruthState = (
        "not_required"
    )

    evidence: list[ToolEvidence] = Field(
        default_factory=list
    )

    memory_updates: list[MemoryFact] = Field(
        default_factory=list
    )

    conversation_state: ConversationState | None = None

    error: str | None = None

    metadata: dict[str, Any] = Field(
        default_factory=dict
    )


CHAT_CONTRACT_VERSION = "134A.v1"
